com.avaloq.tools.ddk.xtext/src/com/avaloq/tools/ddk/xtext/scoping/AbstractPolymorphicScopeProvider.java

```diff
@@ -296,32 +298,38 @@ protected boolean doGlobalCache(final EObject context, final EClass type, final
    */
   protected List<IContainer> getVisibleContainers(final EObject context, final Resource originalResource) { // NOPMD by WTH on 26.01.11 09:26 (NPath
                                                                                                             // complexity...)
-    final Resource ctxRsc = originalResource == null ? context.eResource() : originalResource;
-    if (!(ctxRsc instanceof XtextResource)) {
-      LOGGER.error(MessageFormat.format("Context {0} is not in an Xtext resource: {1}", context, ctxRsc != null ? ctxRsc.getURI() : "null")); //$NON-NLS-1$ //$NON-NLS-2$
+    final Resource resource = originalResource == null ? context.eResource() : originalResource;
+    if (!(resource instanceof XtextResource)) {
+      LOGGER.error(MessageFormat.format("Context {0} is not in an Xtext resource: {1}", context, resource != null ? resource.getURI() : "null")); //$NON-NLS-1$ //$NON-NLS-2$
       throw new IllegalStateException();
     }
-    final XtextResource rsc = (XtextResource) ctxRsc;
-    // Cache these container lists, they're expensive to create
-    URI uri = rsc.getURI();
+    return getVisibleContainers((XtextResource) resource);
+  }
+
+  /**
+   * Return the visible containers given a resource.
+   *
+   * @param resource
+   *          the resource. Must not be {@code null}
+   * @return The list of visible containers.
+   */
+  protected List<IContainer> getVisibleContainers(final XtextResource resource) {
+    URI uri = resource.getURI();
     List<IContainer> result = null;
-    final ResourceCache<URI, List<IContainer>> cache = CacheManager.getInstance().getOrCreateResourceCache("AbstractPolymorphicScopeProvider#visibleContainers", rsc); //$NON-NLS-1$
+    final ResourceCache<URI, List<IContainer>> cache = CacheManager.getInstance().getOrCreateResourceCache("AbstractPolymorphicScopeProvider#visibleContainers", resource); //$NON-NLS-1$
     if (cache != null) {
       result = cache.get(uri);
       if (result != null) {
         return result;
       }
     }
 
-    final EObject ctx = ctxRsc != context.eResource() ? ctxRsc.getContents().get(0) : context;

```

Comment:
The rest of the change is good for me, but I do not understand why we needed this condition or why we can remove it now.

Reply:
I don't understand why we needed the condition either.

The only use that is made of the ctx variable is to get ctx.eResource() in lines 323 and 324. In all cases, the result of ctx.eResource() is ctxRsc (which is now named resource):

- if context.eResource() and originalResource are the same, it's the same
- if originalResource is null, it's replaced by context.eResource(), so they are the same
- if context.eResource() and originalResource are different:
  - ctxRsc == originalResource
  - ctx == originalResource.getContents().get(0)
  - ctx.eResource() == ctxRsc